### ISSUE TITLE
[docs] disconnected foreach statements: cloning outside of a loop

### ISSUE THREAD

**Issue by mikeSimonson** [kalessil/phpinspectionsea#123]
Hi,

First thank you for your great plugin.

I recently stumbled across this warning:
  > Objects should be created outside of a loop and cloned instead.

And I was wondering if you could explain or point to an explanation as to why this is the recommended behaviour. 

Thanks

**Comment by kalessil** [comment]
Thank you @mikeSimonson for reporting and the feedback.

Eah, docs are the weak spot currently. The warning can be disabled, as the inspection has individual setting allowing to do this.

And the root case is following;
```php
foreach ($source as $valueToStore) {
    $databaseEntry = new ModelClass(); // time-stampable
    $databaseEntry->setValue($valueToStore->generate()); // can block for some time
    $databaseEntry->save();
}
```
The entries had differnt created/updated timestamps -> analytics got broken, data are for junk. 
The reliable solution was:
```php
$entry = new ModelClass(); // time-stampable
$entry->setCreatedAt(new \DateTime());
$entry->setUpdatedAt(new \DateTime());

foreach ($source as $valueToStore) {
    $databaseEntry = clone $entry;
    $databaseEntry->setValue($valueToStore->generate()); // can block for some time
    $databaseEntry->save();
}
```

I'm not sure, but perhaps cloning will be even faster in most of cases.

**Comment by mikeSimonson** [comment]
Thanks for the explanation.

BTW I couldn't find the setting to disable it in settings > editor > inspections > EA extended > ...
I couldn't find anything related to cloning.
Is it the "Callable calls in loops termination condition" ?

**Comment by kalessil** [comment]
You are welcome, and the inspection in Php Inspections (...)->Control Flow->Statement could be decoupled from foreach

**Comment by kalessil** [comment]
We ended up disabling the option by default in #132 )

**Comment by rentalhost** [comment]
I don't understand why you disabled this option by default. Seems a good case to refactor. If we have an issue with timestamp (that should be updated, for instance), it should be done on `__clone()` magic method, does not?

I remember that the unique problem that I have with this inspection is because I don't understand what it expected. But _after I tune my english_ I get it and worked good for me. hehe

---

I do some benchmark to check three cases, anyway. I repeat each test 5 times, 100.000 iterations each and calculated the average time, memory and memory peak.

**Case 1:** if you declare `new` in each looping (http://pastebin.com/q7HT1hTY).

> Time: 540.40 ms
> Memory: 2.00 Mb
> Memory Peak; 2.00 Mb

**Case 2:** if you declare `new` outside the looping, but depends of auxiliary methods to update _what constructor should do_ (http://pastebin.com/UxaAu2jL).

My notes: seems that is the best case, besides it should affect the code manutention.

> Time: 447.0 ms (-17.3%)
> Memory: 2.00 Mb
> Memory Peak; 2.00 Mb

**Case 3:** if you declare `new` outside the looping, but depends of auxiliary `__clone()` method to update _what constructor should do_ (http://pastebin.com/8ha6wKJg).

My notes: seems technically the same of case 1. But the major problem here is that my `__clone()` method is an additional method call (which affects the benchmark). If I inline the `__clone()` content (what is not recommended), it will be like case 2.

> Time: 536.60 ms (-0.7%)
> Memory: 2.00 Mb
> Memory Peak; 2.00 Mb

**Comment by kalessil** [comment]
@rentalhost the I'd suggest moving this into a separate inspection. Would you please create a new ticket for this + link your comment in the description?

**Comment by rentalhost** [comment]
@kalessil Which new inspection? I guess that it is related to it or then to #132. But tell me how I should title the new ticket.

**Comment by kalessil** [comment]
@rentalhost : "report using cloning outside of loops"

**Comment by zerkms** [comment]
So, as explained at https://github.com/kalessil/phpinspectionsea/issues/123#issuecomment-276437293 this inspection addresses very specific and narrow case?

Is it not harmful when an inspection is not generic? (or at least covers more than 1 case about one feature of one ORM).

**Comment by kalessil** [comment]
@zerkms  the case was starting point for the inspection.

In general preparing object outside of the loop (constructor and setting it up) is a pretty heavy operation, so cloning the object is faster, [see the comment](https://github.com/kalessil/phpinspectionsea/issues/123#issuecomment-278390132).

**Comment by zerkms** [comment]
@kalessil the difference is minor. While the inspection produces false positives and is not equivalent: you cannot simply substitute a constructor invocation with a clone operation.

I always thought the plugin addresses architectural improvements or mistakes, not arguable micro-optimisations.

**Comment by kalessil** [comment]
@zerkms : correct, plus [this](https://ocramius.github.io/extremely-defensive-php/#/100). And seeding this a bad practice among junors/mids already spotted.

Basically identical argument were the reason for disabling clone suggestion by default. Currently I keep the option on to actually spot pieces of code to look onto and we'll drop the clone related inspection part in near future.
